### DISCUSSION TITLE
padri & antenati comuni

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#12] [Esercizi A-L]
Dato un albero rappresentato come vettore di padri scrivere lo pseudocodice di un algoritmo che trova gli antenati comuni di due vertici $v$ e $w$ in $O(n)$

**Comment by LeoCosta02** [comment]
```C
Antenati_comuni(P : int[n],x: vert , y:vert, Vett1,Vett2){
  Vett1[x],Vett2[y]=True
  
   if ((P[x]==x and P[y]==y))
     return
   
   Antenati_comuni(P,P[x],P[y],Vett1,Vett2)
  
   return
}


Global_antenati(){
P: int[n] //vettore dei padri di un albero
Vett1[n] : vettore booleano che tiene traccia dei nodi già visti da x,inizializzato a F  
Vett2[n]:  vettore booleano che tiene traccia dei nodi già visti da y ,inizializzato a F 
L : list // output 
vert x,y //due vertici dell albero
if(P[x]==x || P[y]==y) return
Antenati_comuni(P,P[x],P[y],Vett1,Vett2)
 for(int i=0 ; i<n; i++){
     if(vett1[i] == true AND  vett2[i]== true)
     L.insert(i)
 }
return L
}
```

**Reply by IonutCicio** [reply]
Ho editato il post facendo quello che volevi fare te con il codice. Appena posso do un'occhiata all'algoritmo, perché pare sbagliato a prima vista...

**Reply by LeoCosta02** [reply]
Per editarlo cosi come devo fare?

**Reply by IonutCicio** [reply]
Questa è la [documentazione ufficiale di GitHub](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) sui code block. Io darei un'occhiata a tutta la [documentazione](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/quickstart-for-writing-on-github), solo per capire come funziona in generale Markdown, è molto utile per tante cose nel mondo software _(documentazione, siti, relazioni etc...)_

**Comment by IonutCicio** [comment]
```rust
fn common_daddies(tree: &[usize], mut x: usize, mut y: usize) -> Vec<usize> {
    let mut daddies = vec![];

    let mut x_daddies = vec![false; tree.len()];

    while x != tree[x] {
        x = tree[x];
        x_daddies[x] = true;
    }

    while y != tree[y] {
        y = tree[y];
        if x_daddies[y] {
            daddies.push(y);
        }
    }

    daddies
}
```

**Comment by LeoCosta02** [comment]
@CuriousCI ho modificato il codice , ora dovrebbe andare

**Comment by CasuFrost** [comment]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/97389762/631cd32c-1eb5-4cad-9cef-09ed0a0bc037)
